---
title: Primitive Types Hacks + More Java 
description: Here is my work for Week 4, which had us do more work with Java's data types!
courses: { csa: {week: 4} }
categories: [C3.0, C3.1, C4.1]
type: tangibles
---

# Overview of this Blog

To recap, this week we were introduced to some of Java's basic types, which can be broken down into primitive types (int, boolean, double, etc.) and non primitive types (strings). As someone who has had coding experience, especially in Java, for several years, none of the information presented was particularly new to me; however, I do think that it is important to make sure that I fully understand these types, as that will be one of the major parts of the exam in May, especially with the FRQs. 

While this blog does include the suggested hacks for this week, I decided to add in a couple more of my personal coding projects for fun, as I generally enjoy coding whenever I have free time. The first couple of cells are the suggested hacks, which are to make a class that demonstrates the use of the primitive and non primitive types in Java and to demonstrate inputs and outputs (multiple). In addition to those, I added a few cells that included my own Java projects, such as a reversed string algorithm and many others that you will see. 

I hope that if anybody happens to see this blog that it will serve as a useful study tool as we begin preparing and reviewing for the exam in May!

# Primitive + Reference Types Demo

In [59]:
import java.util.Scanner;

public class Person {

    private String name; // of course, a string would be best to use here because a string is a series of characters, which is what a name normally is
    private String pair; // a string would best be used for describing someone's pair/partner for the same reason as above
    private int age; // an integer would best be used here because age is typically expressed as a whole number rather than a decimal, hence we use an int instead of a float
    private boolean canDrive; // a boolean is used here because someone can either drive (true) or they cannot (false)
    private float salary; // We could use an integer here, but a salary may not always turn out to be a perfect round number. Floats allow decimals, hence why we are using one here

    public void enterInfo(){


        Scanner sc = new Scanner(System.in); // 
        System.out.println("Enter your full name: ");
        name = sc.nextLine();


        Scanner sc1 = new Scanner(System.in);
        System.out.println("Enter your age: ");
        age = sc1.nextInt();

       
        Scanner sc2 = new Scanner(System.in);
        System.out.println("Can you drive? (y/n): ");
        char driveResponse = sc2.nextLine().charAt(0);

        if(driveResponse == 'y' || driveResponse == 'Y'){
            canDrive = true;
        } else {
            canDrive = false;
        }

        Scanner sc3 = new Scanner(System.in);
        System.out.println("What is your annual salary? ");
        salary = sc3.nextFloat();

        
        Scanner sc4 = new Scanner(System.in);
        System.out.println("Who is your pair in AP CSA?\n");
        pair = sc4.nextLine();
    }


    public static void main(String [] args){

        Person me = new Person();
        me.enterInfo();


        System.out.println("STATS\nName: "+me.name);
        System.out.println("Age: "+ me.age);
        System.out.println("Can Drive: "+me.canDrive);
        System.out.println("Annual Salary: $"+me.salary);
        System.out.println("Pair in AP CSA: "+me.pair);




    }

}

Person.main(null)

Enter your full name: 
Enter your age: 
Can you drive? (y/n): 
What is your annual salary? 
Who is your pair in AP CSA?

STATS
Name: Emaad Mir
Age: 16
Can Drive: true
Annual Salary: $268321.38
Pair in AP CSA: Varalu N


# Type Casting (Widening) Demo 

In [60]:
// Note: Widening goes in this order: byte -> short -> char -> int -> long -> float -> double


public class Widening {
    public static void main(String [] args){
        int myLuckyInt = 19;
        long myLuckyLong = myLuckyInt;
        float myLuckyFloat = myLuckyLong;
        double myLuckyDouble = myLuckyFloat;

        System.out.println("Integer (original): "+myLuckyInt);
        System.out.println("Integer to Long: " +myLuckyLong);
        System.out.println("Long to Float: "+myLuckyFloat);
        System.out.println("Float to Double: "+myLuckyDouble);
    }
}

Widening.main(null);

Integer (original): 19
Integer to Long: 19
Long to Float: 19.0
Float to Double: 19.0


# Type Casting (Narrowing) Demo

In [61]:
// Note: Narrowing goes in this order: double -> float -> long -> int -> char -> short -> byte



public class Narrowing {
    public static void main(String [] args){
        double myDoubleSalary = 213876.1592d;
        float myFloatSalary = (float) myDoubleSalary;
        long myLongSalary = (long) myFloatSalary;
        short myShortSalary = (short) myLongSalary;
        byte myByteSalary = (byte) myShortSalary;

        System.out.println("Double salary value: "+myDoubleSalary);
        System.out.println("Double to float salary value: "+myFloatSalary);
        System.out.println("Float to long salary value: "+myLongSalary);
        System.out.println("Long to short salary value: "+myShortSalary);
        System.out.println("Short to byte salary value: "+myByteSalary);

    }
}

Narrowing.main(null);

Double salary value: 213876.1592
Double to float salary value: 213876.16
Float to long salary value: 213876
Long to short salary value: 17268
Short to byte salary value: 116


# Important Differences to Note about Widening and Narrowing Casting

For both the widening and narrowing process, we convert one data type to another (as seen in both code cells). However, the one thing that separates the two is the fact that widening (going from a smaller type to a larger type size) can be done automatically while narrowing (going from a larger type to a smaller type size) has to be done manually.

In other words, whenever we perform casting on a certain variable with a certain data type (ex. int), if we want to convert it to, say, a double, all we have to do is set the previously defined variable equivalent to a new declared variable that is now a double. This is shown in the first code cell, where I was able to convert myLuckyInt, which was set to 11 (integer), into a double called myLuckyDouble (11.0) without having to explicitly add (double) next to the myLuckyInt.

On the other hand, in the second code cell where narrowing takes place, when I wanted to convert the double mySalary into an integer, I had to perform one more step. Like usual, I defined the double mySalary as a decimal (and added the d at the end to indicate that it is a double). Then, I made a new integer called myWholeSalary, but instead of just being able to convert mySalary into an integer automatically, I had to add (int) in front of mySalary to indicate that I am converting the value into an integer. As a result the double version of salary will print out 213876.15, while the integer version of salary will only print out 213876. 



# Compound Assignment Operator

In [7]:
import java.util.Scanner;

public class ChooseAssignmentOperator{

    static int add(int x, int y){
        return x+=y; //  adds the two numbers together, returns sum
    }

    static int sub(int x, int y){
        return x-=y;  //  subtracts the two numbers, returns difference
    }

    static int multi(int x, int y){
        return x*=y;  //  multiplies the two numbers together, returns product
    }

    static int div(int x, int y){
        if(y == 0){
            System.out.println("Undefined. You cannot divide by 0!"); // if the user's second number is 0, the quotient will just be set to -1, as you can't divide by 0
            return -1;
        }

        return x/=y; // divides two numbers, returns quotient
    }

    static int mod(int x, int y){
        if(y == 0){
            System.out.println("Undefined. You cannot calculate modulus with 0!"); // if the user's second number is 0, the remainder will just be set to -1, since you can't find the remainder with 0
            return -1;
        }

        return x%=y; // finds the remainder from dividing two numbers
    }


    public static void main(String [] args){
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter your first number: ");
        int num1 = scanner.nextInt();

        System.out.println("Enter your second number: ");
        int num2 = scanner.nextInt();


        System.out.printf("Mathematical operations on the numbers %d and %d:\n", num1, num2);


        System.out.println("Sum: "+add(num1, num2));
        System.out.println("Difference: "+sub(num1, num2));
        System.out.println("Product: "+multi(num1, num2));
        System.out.println("Quotient: "+div(num1, num2));
        System.out.println("Remainder: "+mod(num1, num2));


        }

        
    }


ChooseAssignmentOperator.main(null);

Enter your first number: 


Enter your second number: 
Mathematical operations on the numbers 12 and 21:
Sum: 33
Difference: -9
Product: 252
Quotient: 0
Remainder: 12


# Extra Java Programs

## Weather Converter + Above/Below Freezing or Boiling Point of Water

In [42]:
// formula to convert from fahrenheit to celsius is (F-32)*(5/9)
// formula to convert from celsius to fahrenheit is (C*(9/5)) + 32

import java.util.Scanner;

public class WeatherConverter {
    
    private static boolean isBelowFreezing;
    private static boolean isAboveBoiling;

     static double toCelsius(double f){
        return (f-32)*(5.0/9);
    }

     static double toKelvin(double f){
        return (f-32)*(5.0/9) + 273.15;
    }

     static boolean checkFreezing(double f){
        if(f < 32){
            isBelowFreezing = true;
            System.out.println("below the freezing point of water");
        } else{
            System.out.println("above the freezing point of water");
        }
        return isBelowFreezing;
        

        }

     static boolean checkBoiling(double f){
        if(f > 212){
            isAboveBoiling = true;
            System.out.println("above the boiling point of water");
        } else {
            System.out.println("below the boiling point of water");
        }
        return isAboveBoiling;
     }

    public static void main(String [] args){

        Scanner sc = new Scanner(System.in);
        System.out.println("Enter a temperature in degrees Fahrenheit that you would like to convert into degrees Celsius and degrees Kelvin: ");

        double temp = sc.nextDouble();

      
        double celsiusResult = WeatherConverter.toCelsius(temp);
        double kelvinResult = WeatherConverter.toKelvin(temp);
        boolean scale = WeatherConverter.checkFreezing(temp);
        boolean scale1 = WeatherConverter.checkBoiling(temp);

        System.out.printf("%f DEGREES FAHRENHEIT CONVERSIONS:\n", temp);
        System.out.printf("Celsius: %f degrees Celsius\n",celsiusResult);
        System.out.printf("Kelvin: %f degrees Kelvin\n",kelvinResult);
        System.out.printf("Your temperature is %s the freezing point of water.\n", (scale ? "below" : "above"));
        System.out.printf("Your temperature is %s the boiling point of water.\n", (scale1 ? "above" : "below"));




    }


}

WeatherConverter.main(null);

Enter a temperature in degrees Fahrenheit that you would like to convert into degrees Celsius and degrees Kelvin: 
above the freezing point of water
below the boiling point of water
129.000000 DEGREES FAHRENHEIT CONVERSIONS:
Celsius: 53.888889 degrees Celsius
Kelvin: 327.038889 degrees Kelvin
Your temperature is above the freezing point of water.
Your temperature is below the boiling point of water.


## CRUD Operations

In [58]:
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class Crud {

    static ArrayList<Integer> numbers = new ArrayList<>(Arrays.asList(12,29,2,11,39,17));

    static ArrayList<Integer> create(int x){
        numbers.add(x);
        return numbers;
    }

    static ArrayList<Integer> remove(int x){
        numbers.remove((Integer)x);
        return numbers;
    }
    

    public static void main(String [] args){
        Scanner sc = new Scanner(System.in);
               
        while(true){
            System.out.println("Current list of numbers: "+numbers);
            System.out.println("What would you like to do with this list? 0. Exit, 1. Add, 2. Remove");

            int response = sc.nextInt();
            if(response == 0){
                break;
            } else if (response == 1){
                System.out.println("What number would you like to add to the list?");
                int numToAdd = sc.nextInt();
                create(numToAdd);
            } else if (response == 2){
                System.out.println("What number would you like to remove from the list?");
                int numToRemove = sc.nextInt();
                if (numbers.contains(numToRemove)) {
                    remove(numToRemove);
                } else {
                    System.out.println("Number not found in the list.");
                }

            } else {
                System.out.println("Invalid choice. Please enter 0, 1, or 2.");
            }
        }


        System.out.println("Final numbers: " + numbers);
    }



}


Crud.main(null);

Current list of numbers: [12, 29, 2, 11, 39, 17]
What would you like to do with this list? 0. Exit, 1. Add, 2. Remove
What number would you like to remove from the list?
Current list of numbers: [29, 2, 11, 39, 17]
What would you like to do with this list? 0. Exit, 1. Add, 2. Remove
What number would you like to remove from the list?
Current list of numbers: [29, 2, 11, 39]
What would you like to do with this list? 0. Exit, 1. Add, 2. Remove
What number would you like to add to the list?
Current list of numbers: [29, 2, 11, 39, 1020]
What would you like to do with this list? 0. Exit, 1. Add, 2. Remove
Final numbers: [29, 2, 11, 39, 1020]
